winml/adapter/winml_adapter_model.cpp

```diff
@@ -777,6 +790,129 @@ ORT_API_STATUS_IMPL(winmla::OperatorGetOutputName, _In_ const char* const op_typ
+    auto old_name = other_input.name();
+    *other_input.mutable_name() = second_model_prefix + old_name;
+
+    auto found_it = std::find_if(input_names, input_names + num_linkages,
+                                 [old_name](auto& name) { return std::strcmp(name, old_name.c_str()) == 0; });
+    bool is_linked = found_it != (input_names + num_linkages);  // figure out if other_input.name() exists in the output_names mapped
+    if (!is_linked) {
+      auto& input = *first_model_proto->mutable_graph()->add_input();
+      input = std::move(other_input);
+    }
+  }
+
+  // add all initializers
+  for (int i = 0; i < second_model_proto->graph().initializer_size(); i++) {
+    auto& other_initializer = *second_model_proto->mutable_graph()->mutable_initializer(i);
+    *other_initializer.mutable_name() = second_model_prefix + other_initializer.name();
+    auto& initializer = *first_model_proto->mutable_graph()->add_initializer();
+    initializer = std::move(other_initializer);
+  }
+
+  // add all nodes
+  for (int i = 0; i < second_model_proto->graph().node_size(); i++) {
+    auto& other_node = *second_model_proto->mutable_graph()->mutable_node(i);
+    if (0 != strcmp(other_node.name().c_str(), "")) {
+      *other_node.mutable_name() = second_model_prefix + other_node.name();
+    }
+    for (int j = 0; j < other_node.input_size(); j++) {
+      *other_node.mutable_input(j) = second_model_prefix + other_node.input(j);
+    }
+    for (int j = 0; j < other_node.output_size(); j++) {
+      *other_node.mutable_output(j) = second_model_prefix + other_node.output(j);
+    }
+    auto& node = *first_model_proto->mutable_graph()->add_node();
+    node = std::move(other_node);
+  }
+
+  int64_t opset = 7;
+  for (int i = 0; i < second_model_proto->opset_import_size(); i++) {
+    auto mutable_opset_import = second_model_proto->mutable_opset_import(i);
+    auto domain = mutable_opset_import->has_domain() ? mutable_opset_import->domain() : std::string("");
+    auto version = mutable_opset_import->version();
+
+    // does the domain exist in the first model?
+    auto found_it = std::find_if(first_model_proto->mutable_opset_import()->begin(), first_model_proto->mutable_opset_import()->end(),
+                                 [&domain](auto& mutable_opset_import) {
+                                    
+                                    auto first_model_domain = mutable_opset_import.has_domain() ? mutable_opset_import.domain() : std::string("");
+                                    return 0 == strcmp(first_model_domain.c_str(), domain.c_str());
+                                 });
+    if (found_it != first_model_proto->mutable_opset_import()->end()) {
+      found_it->set_version(std::max(found_it->version(), version));
+      if (0 == strcmp(domain.c_str(), "")) {
+        opset = found_it->version();
+      }
+    }
+  }
+
+  // add identity ops to rename all of the first model outputs to secondmodel inputs with prefix for each linkage
```

Comment:
can you just rename the first model outputs? does identity add much overhead?

Reply:
Identity doesnt add overhead and gets optimized out.
Also I cant rename the second model easily, as i would have to update many nodes that reference it.
I cant rename the first easily, because i would have to update potential nodes (not common) that reference a model output.

Identity was an easy way to rename.